### PR TITLE
FIX let CompositeField subclasses render themselves

### LINKED ISSUE #37: CompositeField_holder does too much
It implements the entire field, which is what the `Field` call is for. This results in subclasses (not to be confused with children) never having a chance to render themselves correctly.

e.g. SelectionGroup is supposed to render a list, as evident by the SelectionGroup template. However, when called in a template it will (as with all form fields) render in it's `_holder` template, ultimately resulting in no `ul` or `li` elements ever being rendered (by nature of `Field` being cut out of the holder template).

This in turn causes unit test failures in `SilverStripe\Forms\Tests\SelectionGroupTest`.

### PR DESCRIPTION
**Pull request by NightJar**
resolves #37

### CODE REVIEW

**Comment by raissanorth** [comment]
I have implemented these changes in a user defined form created via the CMS.

**HTML before:**

```
   <div id="EditableFormStep_147bc" class="form-step field CompositeField userformsstep viewed" data-title="First Page" aria-hidden="false" style="">
		<div id="EditableTextField_b9c00" class="field form-group text">
    <label class="control-label" for="UserForm_Form_56_EditableTextField_b9c00" id="title-UserForm_Form_56_EditableTextField_b9c00">
        Text Field before Group
    </label>
    <input type="text" name="EditableTextField_b9c00" id="UserForm_Form_56_EditableTextField_b9c00" class="text form-control" aria-labelledby="title-UserForm_Form_56_EditableTextField_b9c00" style="background-image: url(&quot;data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAABHklEQVQ4EaVTO26DQBD1ohQWaS2lg9JybZ+AK7hNwx2oIoVf4UPQ0Lj1FdKktevIpel8AKNUkDcWMxpgSaIEaTVv3sx7uztiTdu2s/98DywOw3Dued4Who/M2aIx5lZV1aEsy0+qiwHELyi+Ytl0PQ69SxAxkWIA4RMRTdNsKE59juMcuZd6xIAFeZ6fGCdJ8kY4y7KAuTRNGd7jyEBXsdOPE3a0QGPsniOnnYMO67LgSQN9T41F2QGrQRRFCwyzoIF2qyBuKKbcOgPXdVeY9rMWgNsjf9ccYesJhk3f5dYT1HX9gR0LLQR30TnjkUEcx2uIuS4RnI+aj6sJR0AM8AaumPaM/rRehyWhXqbFAA9kh3/8/NvHxAYGAsZ/il8IalkCLBfNVAAAAABJRU5ErkJggg==&quot;); background-repeat: no-repeat; background-attachment: scroll; background-size: 16px 18px; background-position: 98% 50%;">
</div>
		<fieldset class="CompositeField field CompositeField userformsgroup ">
```
```
			<fieldset class="CompositeField field CompositeField userformsgroup ">
```

**HTML after:**
```
    <div id="EditableFormStep_147bc" class="form-step field CompositeField userformsstep viewed" data-title="First Page" aria-hidden="false" style="">
		<div id="EditableTextField_b9c00" class="field form-group text">
    <label class="control-label" for="UserForm_Form_56_EditableTextField_b9c00" id="title-UserForm_Form_56_EditableTextField_b9c00">
        Text Field before Group 
    </label>
    <input type="text" name="EditableTextField_b9c00" id="UserForm_Form_56_EditableTextField_b9c00" class="text form-control" aria-labelledby="title-UserForm_Form_56_EditableTextField_b9c00" style="background-image: url(&quot;data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAABHklEQVQ4EaVTO26DQBD1ohQWaS2lg9JybZ+AK7hNwx2oIoVf4UPQ0Lj1FdKktevIpel8AKNUkDcWMxpgSaIEaTVv3sx7uztiTdu2s/98DywOw3Dued4Who/M2aIx5lZV1aEsy0+qiwHELyi+Ytl0PQ69SxAxkWIA4RMRTdNsKE59juMcuZd6xIAFeZ6fGCdJ8kY4y7KAuTRNGd7jyEBXsdOPE3a0QGPsniOnnYMO67LgSQN9T41F2QGrQRRFCwyzoIF2qyBuKKbcOgPXdVeY9rMWgNsjf9ccYesJhk3f5dYT1HX9gR0LLQR30TnjkUEcx2uIuS4RnI+aj6sJR0AM8AaumPaM/rRehyWhXqbFAA9kh3/8/NvHxAYGAsZ/il8IalkCLBfNVAAAAABJRU5ErkJggg==&quot;); background-repeat: no-repeat; background-attachment: scroll; background-size: 16px 18px; background-position: 98% 50%; cursor: auto;">
</div>
		<fieldset class="CompositeField field CompositeField userformsgroup ">
```
```
			<fieldset class="CompositeField field CompositeField userformsgroup ">
```

**Screenshot before:**
image

**Screenshot after:**
image

**Comment by robbieaverill** [comment]
What am I looking at in terms of the differences?

**Comment by raissanorth** [comment]
I run a diff and there were _no_ differences.

**Comment by robbieaverill** [comment]
What's the point of the pull request then? :D

**Comment by robbieaverill** [comment]
Bump

**Comment by NightJar** [comment]
image

**Comment by NightJar** [comment]
Without patch:
image

Note the critical lack of radio buttons

**Comment by robbieaverill** [comment]
Are they the same data examples? I don't see the correlation

**Comment by NightJar** [comment]
Yes, sorry. Forgot the corresponding code part:

```
<?php

use CWP\CWP\PageTypes\BasePageController;
use SilverStripe\Forms\Form;
use SilverStripe\Forms\FieldList;
use SilverStripe\Forms\TextField;
use SilverStripe\Forms\EmailField;
use SilverStripe\Forms\FormAction;
use SilverStripe\Forms\SelectionGroup;
use SilverStripe\Forms\SelectionGroup_Item;

class PageController extends BasePageController
{
    public function Form()
    {
        return Form::create(
            $controller = $this,
            $factoryFunction = __FUNCTION__,
            $fields = FieldList::create(
                SelectionGroup::create(
                    $name = 'sgtest',
                    $items = [
                        SelectionGroup_Item::create(
                            $value = 'one',
                            $fields = [
                                TextField::create('name1', 'Name'),
                                EmailField::create('email1', 'Email'),
                            ],
                            $title = 'First'
                        ),
                        SelectionGroup_Item::create(
                            $value = 'two',
                            $fields = [
                                TextField::create('name2', 'Name'),
                                EmailField::create('email2', 'Email'),
                            ],
                            $title = 'Second'
                        ),
                    ],
                    $value = null
                )
            ),
            $actions = FieldList::create(
                FormAction::create('doThing', 'OK')
            ),
            $requiredFields = null
        );
    }
}
```

**Comment by robbieaverill** [comment]
Ok cool, the markup looks fine and the screenshots look improved. 👍

File: templates/SilverStripe/Forms/CompositeField.ss [hunk 1]
```diff
@@ -1,14 +1,9 @@
...
+        <div class="column-{$ColumnCount.ATT} {$FirstLast.ATT}">
             {$Field}
-        <% end_if %>
-    <% end_loop %>
-</$Tag>
+        </div>
+    <% else %>
+        {$Field}
+    <% end_if %>
+<% end_loop %>
```
Comment (robbieaverill) [inline templates/SilverStripe/Forms/CompositeField.ss]: Sorry, premature approval.

This template is the same now as the one in framework, but I note that the template in the cms-forms theme for admin has markup that's more similar to the original. Can you please investigate which of those two is more appropriate, and if it's framework then delete this template and let it inherit through, if it's CMS forms then adjust this template to suit it?

The purpose of overriding these templates in the theme is (A) to improve accessibility (some of which has been done in SS4 core now) and (B) to style the form fields for Bootstrap 3. This template originally had some bootstrap classes, but doesn't in the proposed version so probably won't provide any value any more.

File: templates/SilverStripe/Forms/SelectionGroup.ss [hunk 2]
```diff
@@ -13,7 +13,7 @@
     <ul class="{$ExtraClass.ATT}">
         <% loop $FieldSet %>
             <li {$Selected}>
-                {$RadioButton} {$RadioLabel}
+                <label for="{$Up.ID}_{$Pos}">{$RadioButton} {$RadioLabel}</label>
```
Comment (robbieaverill) [inline templates/SilverStripe/Forms/SelectionGroup.ss]: If `$RadioButton` renders an input and it's inside the label element, then we don't need the `for` attribute.

See `SelectionGroup.ss` in framework: `<label>{$RadioButton} {$RadioLabel}</label>`

**Pull request merged**